src/main/scala/org/allenai/plugins/DeployPlugin.scala

```diff
@@ -236,64 +237,170 @@ object DeployPlugin extends AutoPlugin {
+    val deploys: Iterable[Future[Try[Unit]]] = hosts flatMap { hostConfig =>
+      val deployHost = hostConfig.host
+      val baseDeployDir = deployConfig.baseDirectory
 
-          Seq("ssh", "-i", keyFile, "-l", hostConfig.sshUser)
-        }
-        val deployHost = hostConfig.host
-        val deployDirectory = hostConfig.directory
-
-        val rsyncDirs = deployDirs.value map (name => s"--include=/$name")
-        val rsyncCommand = Seq.concat(
-          Seq("rsync", "-vcrtzP", "--rsh=" + sshCommand.mkString(" ")),
-          rsyncDirs,
-          Seq(
-            "--exclude=/*",
-            "--delete",
-            universalStagingDir.getPath + "/",
-            deployHost + ":" + deployDirectory
-          )
-        )
+      // Command to pass to rsync's "rsh" flag, and to use as the base of our ssh operations.
+      val sshCommand = {
+        val keyFile = sys.env.getOrElse("AWS_PEM_FILE", {
+          throw new IllegalStateException(s"Environment variable 'AWS_PEM_FILE' is undefined")
+        })
+        require(keyFile.nonEmpty, "Environment variable 'AWS_PEM_FILE' is empty")
 
-        // Shell-friendly version of rsync command, with rsh value quoted.
-        val quotedRsync = rsyncCommand.patch(
-          2, Seq("--rsh=" + sshCommand.mkString("\"", " ", "\"")), 1
-        ).mkString(" ")
-
-        // Now, ssh to the remote host and run the restart script.
-        val restartScript = deployDirectory + "/" + hostConfig.startupScript
-        val restartArgs = if (hostConfig.startupArgs.nonEmpty) {
-          "--" +: hostConfig.startupArgs
-        } else {
-          Seq()
-        }
-
-        val restartCommand = Seq.concat(
-          sshCommand,
-          Seq(deployHost, restartScript, "restart"),
-          restartArgs
-        )
+        Seq("ssh", "-i", keyFile, "-l", hostConfig.sshUser)
+      }
 
-        // Run commands within a Try so we can collect errors using Future.fold.
-        // TODO(danm): Use ProcessIO to capture and report the actual stderr of failures here.
-        Try {
-          log.info("Running " + quotedRsync + " . . .")
-          if (Process(rsyncCommand).! != 0) {
-            sys.error(s"Error running rsync to host '$deployHost'.")
-          }
+      val replicas = deployConfig.replicaOverridesByHost(hostConfig)
+      val numReplicas = replicas.length
+
+      // First, stop and remove any replicas on the remote that won't be restarted
+      // as part of the deploy.
+      val (deployParent, deployTarget) = baseDeployDir.splitAt(baseDeployDir.lastIndexOf('/'))
+      // Strip leading / from deployTarget so we can match names on it.
+      val namePattern = deployTarget.tail
+      // Build a regex matching the replicas we will eventually update / restart as part of this
+      // process, and therefore DON'T want to stop and remove.
+      val restartedPattern = if (numReplicas == 1) {
+        namePattern
+      } else {
+        s"$namePattern-[1-$numReplicas]"
```

Comment:
So, I don't think we want *all* deploys to include a replica number - only those that deploy multiple times to the same host.

This might be too complicated to try to write logic for. Specifying all of the directories seems better than having to worry about which items are needing to be restarted or not.